lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -381,88 +412,283 @@ void GrandCentralSignalMappingsPass::runOnOperation() {
+      });
+  bool allAnalysesPreserved =
+      llvm::all_of(results, [](auto &r) { return r.allAnalysesPreserved; });
+
+  // If this is a subcircuit, then emit JSON information necessary to drive
+  // SiFive tools.
+  // Otherwise handle any forced input ports and emit updated mappings.
+  if (isSubCircuit) {
+    std::string jsonString;
+    llvm::raw_string_ostream jsonStream(jsonString);
+    json::OStream j(jsonStream, 2);
+    j.object([&] {
+      j.attributeObject("vendor", [&]() {
+        j.attributeObject("vcs", [&]() {
+          j.attributeArray("vsrcs", [&]() {
+            for (FModuleOp module : circuit.body().getOps<FModuleOp>()) {
+              SmallVector<char> file(outputFilename.begin(),
+                                     outputFilename.end());
+              llvm::sys::fs::make_absolute(file);
+              llvm::sys::path::append(file, Twine(module.moduleName()) + ".sv");
+              j.value(file);
+            }
+            for (FExtModuleOp ext : extmodules.vsrc) {
+              SmallVector<char> file(outputFilename.begin(),
+                                     outputFilename.end());
+              llvm::sys::fs::make_absolute(file);
+              llvm::sys::path::append(file, Twine(ext.moduleName()) + ".sv");
+              j.value(file);
+            }
+          });
+        });
+        j.attributeObject("verilator", [&]() {
+          j.attributeArray("error", [&]() {
+            j.value("force statement is not supported in verilator");
+          });
+        });
+      });
+      j.attributeArray("remove_vsrcs", []() {});
+      j.attributeArray("vsrcs", []() {});
+      j.attributeArray("load_jsons", [&]() {
+        for (FExtModuleOp extModule : extmodules.json)
+          j.value((Twine(extModule.moduleName()) + ".json").str());
+      });
+    });
+    auto b = OpBuilder::atBlockEnd(circuit.getBody());
+    auto jsonOp = b.create<sv::VerbatimOp>(b.getUnknownLoc(), jsonString);
+    jsonOp->setAttr(
+        "output_file",
+        hw::OutputFileAttr::getFromFilename(
+            b.getContext(), Twine(circuitPackage) + ".subcircuit.json", true));
+  } else {
+    // Remote side, main circuit
+
+    // 1) Sanity checking
+    // 2) Fixup ports that are driven from subcircuit
+    // 3) Generate new remote references
+    //    For now, just drive/probe using `module.name` XMR's,
+    //    it's much simpler than emitting hierarchical paths
+    //    from the top module/DUT, and works for now.
+    // 4) Emit updated mappings as SignalDriverAnnotations
+    //    for consumption when processing the subcircuit
+    //    (as replacements for the original annotations)
+
+    // Helper to instantiate module namespaces as-needed
+    DenseMap<FModuleOp, ModuleNamespace> moduleNamespaces;
+    auto getModuleNamespace =
+        [&moduleNamespaces](FModuleOp module) -> ModuleNamespace & {
+      return moduleNamespaces.try_emplace(module, module).first->second;
+    };
```

Comment:
👍 for a function to lazily build a namespace. This type of thing happens so often that eventually adding a utility would be great.